### ISSUE TITLE
say() does not prevent unicode character truncation

### ISSUE THREAD

**Issue by half-duplex** [sopel-irc/sopel#2314]
### Requested Feature

When Sopel sends a message that results in an IRC line longer than the limit, the message is sometimes truncated in the middle of a unicode character.

Example:
`< mal> .exec bot.say("♥"*200)`
`< Sopel> ♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥??`

I've marked this as a feature request since it is technically unsurprising (IRC) and doesn't really violate any promises in the [documentation](https://preview.sopel.chat/docs/bot.html#sopel.bot.SopelWrapper.say), but it may still be preferable to truncate at a true character boundary if not a word boundary.

The same behavior is observed with `max_messages=1`, but not when `truncation` or `trailing` are used.

Last tested on 17bb5b9f1ba2f8d45be537977fd80e6349a251e7. Note that the byte location of the truncation is dependent on the length of various things including the bot's nick.

**Comment by dgw** [comment]
As of e5bcd1fbfbc781eee990b47c0236f0772ba13124, still possible to reproduce, but I don't think the problem is where truncation occurs.

Compare the raw log (top) vs. what was received by my client (The Lounge, bottom):

```
# default.raw.log
2023-05-02 17:56:12,638 >>      'PRIVMSG #dgw :♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥\r\n'

# The Lounge web UI
17:56:13 <SopelTest2> ♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥♥�
```

While I don't have raw logs in TL, it certainly looks like this is more likely either 1) a bug in how Sopel calculates the safe length, or 2) the result of an incorrect assumption/guess at the safe length when not all required fields for the calculation are available.

That is, the bytestream is correctly truncated at a character boundary, but then later truncated further by the IRCd. What is transmitted to other clients on the network is either 2.33 or 2.66 bytes shorter than the original line sent by Sopel, meaning the IRCd lopped off either 7 or 8 bytes beyond what Sopel expected when it truncated the message before sending.

I will go over the `get_sendable_message()` code, which I last touched in #2310, not long before this issue was filed.

**Comment by SnoopJ** [comment]
### tl;dr — "what about Unicode edge cases?"

The simple fix for the given test-case is to align the truncation to codepoint boundaries (i.e. don't split multi-byte codepoints), but Unicode-brain won't let me get away without mentioning that this issue also applies to multi-codepoint sequences like ones that use a [zero-width joiner](https://en.wikipedia.org/wiki/Zero-width_joiner), combining characters, or variation selectors, and the simple fix will still run into trouble for those cases. In the interest of being specific, here are some examples: 

```python
multi_codepoint_graphemes = [
    "ണ്‍",  # U+0D23 U+0D4D U+200D
    "👩🏿‍⚕️",  # U+1F469 U+1F3FF U+200D U+2695 U+FE0F
    "🏴‍☠️",  # U+1F3F4 U+200d U+2620 U+FE0F
    "ç",    # U+0063 U+0327
]
```

To handle those cases, Sopel would have to know about Unicode's rules for [boundaries between "user-perceived characters"](https://unicode.org/reports/tr29/#Grapheme_Cluster_Boundaries), which might be kind of a big bite for Sopel to take. While pondering on IRC how to find those boundaries without using `PyICU` (which requires compilation and is inappropriate for Sopel), I _did_ discover a nice pure-Python implementation of the same rules: [`grapheme`](https://pypi.org/project/grapheme/).

The edge I'm talking about probably deserves its own issue if it's not squarely in the "Sopel shouldn't do that" category :sweat_smile:

**Comment by dgw** [comment]
99% sure I tracked it down. Need to do a rewrite of this mess:

https://github.com/sopel-irc/sopel/blob/e5bcd1fbfbc781eee990b47c0236f0772ba13124/sopel/irc/__init__.py#L780-L794

That seems to miss actually truncating the `text` to `safe_text_length()` in some cases. Skipping the truncation/trailing logic & just doing `text, excess = tools.get_sendable_message(text, safe_length)` appears to make the same `bot.say('♥'*150)` call that was previously truncated by the IRCd on my test bot, send a shorter message that is not truncated further.

I'll work on that rewrite now.

@SnoopJ You should open a new issue for handling grapheme boundaries instead of just character boundaries.
